C++ API parity: at::Tensor::requires_grad_

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26568 C++ API parity: at::Tensor::register_hook
* **#26332 C++ API parity: at::Tensor::requires_grad_**



Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

pbelevich committed:
C++ API parity: at::Tensor::requires_grad_

yf225 commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -1,6 +1,7 @@
 #include <c10/util/Optional.h>
 #include <torch/csrc/autograd/VariableTypeUtils.h>
 #include <torch/csrc/utils/memory.h>
+#include <torch/csrc/autograd/utils/python_error_messages.h>
```
It would be great to rename `torch/csrc/autograd/utils/python_error_messages.h` to `torch/csrc/autograd/utils/error_messages.h`, since it's not only used in the Python API now.

pbelevich replied:
done

yf225 reviewed (commented):
Overall looks awesome! I left a minor comment.

eellison commented:
This will break things in `JIT` because we have a pass that analyzes whether or not a Tensor has a gradient or not

yf225 commented:
@eellison Do you mind elaborating more on the use cases that this will break? The `at::Tensor::requires_grad_` API is supposed to be used from the C++ frontend, which I think shouldn't break JIT?

eellison commented:
> @eellison Do you mind elaborating more on the use cases that this will break? The `at::Tensor::requires_grad_` API is supposed to be used from the C++ frontend, which I think shouldn't break JIT?



> @eellison Do you mind elaborating more on the use cases that this will break? The `at::Tensor::requires_grad_` API is supposed to be used from the C++ frontend, which I think shouldn't break JIT?

All of the functions in `native_functions.yaml` are exposed in JIT

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

yf225 reviewed (approved):
Thanks a lot @pbelevich !

yf225 commented:
@eellison 

> All of the functions in `native_functions.yaml` are exposed in JIT. We would like to land this in JIT as well just have to update our analysis pass before this lands.

Do you have an example PR for adding support for this in JIT? We can work on that as well to move faster on this API.

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

eellison commented:
> @eellison
> 
> > All of the functions in `native_functions.yaml` are exposed in JIT. We would like to land this in JIT as well just have to update our analysis pass before this lands.
> 
> Do you have an example PR for adding support for this in JIT? We can work on that as well to move faster on this API.

It's complicated enough that it's probably better if someone in JIT does it. cc @bwasti :) 

the logic is similar to https://github.com/pytorch/pytorch/blob/2f0b987ba1ab17828a87e78ec40ed1264968d276/torch/csrc/jit/passes/shape_analysis.cpp#L118 where any value that potentially aliases requires_grad_(True or non-constant) we would need to see the grad conservatively on

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

yf225 commented:
@eellison @bwasti Do you have suggestions on how we should move forward on this?

eellison commented:
@bwasti are you still working on a schema annotation to mark side effects ?

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"


Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

yf225 commented:
@eellison Bumping this again - what should be the next steps to take?

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"




Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"




Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich commented:
@bwasti showed me an example of c10 registration and I added requires_grad_ to torch/csrc/jit/register_prim_ops.cpp

yf225 reviewed (approved):
Thanks a lot @pbelevich!

eellison commented:
> @bwasti showed me an example of c10 registration and I added requires_grad_ to torch/csrc/jit/register_prim_ops.cpp

Could you remove the register_prim_ops implementation ? Those are for registering ops that bound to the torch c++ library.

bwasti commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -738,6 +738,15 @@ RegisterOperators reg(
            return 0;
          },
          aliasAnalysisConservative()),
+     Operator(
```
I believe the only change required is the one in this file

bwasti replied:
err, I take that back -- I believe you've also enabled it in the C++ API

bwasti reviewed (approved):
assuming all tests pass

pbelevich commented:
@eellison @bwasti I'm very confused with your comments. Should I change anything or not?

bwasti commented:
@pbelevich edited my comment

bwasti commented:
@eellison , this one wouldn't be exposed correctly. note the "conservative" annotation in the prim_ops registration.

eellison commented:
> @eellison , this one wouldn't be exposed correctly. note the "conservative" annotation in the prim_ops registration.

Do you know that the register_prim_ops schema is being matched to before the native_functions one?

eellison commented:
> @eellison that seems to be how it is used in the file,
> https://bddppq.github.io/codebrowser/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp.html#2823
> 
> I don't know if bugs have crept in -- the c10 dispatch code is extremely hard to understand. I believe, if there is an issue, there will be a runtime error
> https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/core/dispatch/Dispatcher.cpp.html#63

From looking at the code I would suspect this should raise, since it's the same schema with a different options. If doesn't than i would think that's a bug. cc @smessmer

smessmer reviewed (approved):
lgtm

eellison commented:
> lgtm

@smessmer why is it not a bug that https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/core/dispatch/Dispatcher.cpp.html#63 didn't fire?

smessmer reviewed (changes requested):
oh wait I'm a bit confused here. The entry in `native_functions.yaml` should already create a jit op for this in `register_aten_ops.cpp`. Why is the one in `register_prim_ops.cpp` needed?

@eellison This doesn't crash because `register_prim_ops.cpp` isn't the c10 operator library, that's a shortcut directly to jit which should only be used if absolutely needed.

pbelevich commented:
@smessmer @bwasti can you please reach a consensus?

smessmer reviewed (approved):
withdrawing my concerns with recent changes

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"




Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: at::Tensor::requires_grad_"




Differential Revision: [D17427575](https://our.internmc.facebook.com/intern/diff/D17427575)

[ghstack-poisoned]

resistor commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -111,6 +112,15 @@ int64_t _version(const Tensor & self) {
+Tensor& requires_grad_(Tensor& self, bool _requires_grad) {
+  if (!self.is_leaf() && !_requires_grad) {
+    throw std::runtime_error(
+      autograd::utils::requires_grad_leaf_error(requires_grad)
```
This is a bug.  `requires_grad` is a function, `_requires_grad` is the boolean.

resistor replied:
@pbelevich

yf225 replied:
Fixed in https://github.com/pytorch/pytorch/commit/1e3e1f5bf9745d3c82dc0834064e443d6d142785

wanchaol commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -734,6 +734,15 @@ RegisterOperators reg(
          },
          aliasAnalysisConservative()),
+     Operator(
+         "aten::requires_grad_(Tensor(a!) self, bool _requires_grad=True) -> Tensor(a!)",
```
@pbelevich hey any reason we have this schema with `_requires_grad` instead of `requires_grad`? This is creating discrepancy between the jit api and the python api as it take `requires_grad`..
